File: 程序员面试金典/8.1/waysToStep.java
class Solution {
    public int waysToStep(int n) {
        if (n <= 2)
            return n;
        else if (n == 3)
            return 4;
        int a = 1, b = 2, c = 4;
        for (int i = 4; i <= n; ++i) {
            int temp = ((a + b) % 1000000007 + c) % 1000000007;
            a = b;
            b = c;
            c = temp;
        }
        return c;
    }
}

File: 程序员面试金典/8.1/waysToStep.py
class Solution:
    def waysToStep(self, n: int) -> int:
        if n <= 2:
            return n
        elif n == 3:
            return 4
        dp = [0] * (n + 1)
        dp[1:4] = [1, 2, 4]
        for i in range(4, n + 1):
            dp[i] = (dp[i - 1] + dp[i - 2] + dp[i - 3]) % 1000000007
        return dp[n]
